### DISCUSSION TITLE
zh/guides/heatmaps/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9662] [Docs]
# zh/guides/heatmaps/

利用Ultralytics YOLOv8  热图实现高级数据可视化

https://docs.ultralytics.com/zh/guides/heatmaps/

**Comment by nzbanana** [comment]
I'm having some problems with the script.
(yolo) D:\Vscode\workspace\ultralytics-240618>python D:\Vscode\workspace\ultralytics-240618\heat_map.py.
None

0: 384x640 1 person, 97.5ms
Speed: 3.6ms preprocess, 97.5ms inference, 55.3ms postprocess per image at shape (1, 3, 384, 640)
Traceback (most recent call last):
  File "D:\Vscode\workspace\ultralytics-240618\heat_map.py", line 28, in <module>
    im0 = heatmap_obj.generate_heatmap(im0, tracks)
  File "D:\Vscode\workspace\ultralytics-240618\ultralytics\solutions\heatmap.py", line 138, in generate_heatmap
    self.extract_results(tracks)
  File "D:\Vscode\workspace\ultralytics-240618\ultralytics\solutions\heatmap.py", line 119, in extract_results
    self.track_ids = tracks[0].boxes.id.int().cpu().tolist()
AttributeError: 'NoneType' object has no attribute 'int'

**Reply by glenn-jocher** [reply]
@nzbanana hi there!

Thank you for reaching out and providing the error details. It looks like the issue arises because the `tracks` object is `None`, leading to the `AttributeError`.

To help us investigate further, could you please provide a minimum reproducible example of your script? This will allow us to replicate the issue on our end. You can refer to our guide on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example). 

Additionally, please ensure that you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using the following commands:

```bash
pip install --upgrade torch
pip install --upgrade ultralytics
```

Here's a quick checklist to help debug the issue:
1. **Check if `tracks` is `None`:** Ensure that the `model.track()` method is returning valid tracks.
2. **Verify the video source:** Make sure the video file is correctly read and processed.
3. **Update packages:** Ensure you are using the latest versions of `torch` and `ultralytics`.

Here's a modified snippet to help you debug:

```python
import cv2
from ultralytics import YOLO, solutions

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Video writer
video_writer = cv2.VideoWriter("heatmap_output.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

# Init heatmap
heatmap_obj = solutions.Heatmap(
    colormap=cv2.COLORMAP_PARULA,
    view_img=True,
    shape="circle",
    classes_names=model.names,
)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break
    
    tracks = model.track(im0, persist=True, show=False)
    if tracks is None:
        print("No tracks found.")
        continue

    im0 = heatmap_obj.generate_heatmap(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

This snippet includes a check to ensure `tracks` is not `None` before proceeding with the heatmap generation.

Please try these steps and let us know if the issue persists. We're here to help! 😊

**Comment by fangyuZ1025** [comment]
That just track task , if  i want do some detect, classify,segement task? how can i modify the code?

**Reply by glenn-jocher** [reply]
@fangyuZ1025 to perform detection, classification, or segmentation tasks, you can modify the code by loading the corresponding YOLO11 model (e.g., `yolo11n.pt` for detection, `yolo11n-cls.pt` for classification, or `yolo11n-seg.pt` for segmentation) and applying the desired mode. Refer to the [Ultralytics Docs](https://docs.ultralytics.com/models/yolo11) for detailed examples and configurations. Let me know if you need further clarification!

**Reply by fangyuZ1025** [reply]
Sorry, I think I didn't express what I meant. I am working on the heat map of yolov8 to explain the interpretability of the model. This code is doing the heat map of the video tracking task, and I want to implement my image heat map feature detection based on it. I wonder if there is any relevant code. Thanks again for your reply

**Reply by glenn-jocher** [reply]
Thank you for clarifying! To implement image-based heatmaps for feature detection, you can adapt the existing `Heatmap` class by modifying the input handling to process static images instead of video streams. You can leverage the `generate_heatmap` method to overlay heatmaps on single images. For further guidance, refer to the [`heatmap.py`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/solutions/heatmap.py) implementation. Let us know if you need further assistance!

**Comment by haiyuou** [comment]
Is there a heatmap module compatible with YOLO v11 for image detection?

**Reply by haiyuou** [reply]
I am a beginner from China. If you are willing to answer my questions, please contact me via email. Thank you very much.    <1636705028@qq.com>

**Reply by glenn-jocher** [reply]
@haiyuou thanks for your interest in YOLO11! The current heatmap implementation is designed for video tracking applications, but you can adapt it for single image detection by using the same [Heatmap solution](https://docs.ultralytics.com/guides/heatmaps/) with static images. Please continue the discussion here on GitHub rather than via email so the entire community can benefit from the solutions.
